kube-core/src/schema/transform_one_of.rs

```diff
@@ -0,0 +1,181 @@
+
+use crate::schema::{Schema, SchemaObject, SubschemaValidation};
+
+/// Merge oneOf subschema enums and consts into a schema level enum.
+///
+/// Used for correcting the schema for tagged enums with unit variants.
+///
+/// NOTE: Subschema descriptions are lost when they are combined into a single
+/// enum of the same type.
+///
+/// This will return early without modifications unless:
+/// - There are `oneOf` subschemas (not empty).
+/// - Each subschema contains an `enum` or `const`.
+///
+/// Subschemas must define a type, and they must be the same for all.
+///
+/// NOTE: This should work regardless of whether other hoisting has been
+/// performed or not.
+#[derive(Debug, Clone)]
+pub struct OneOfSchemaRewriter;
+
+impl Transform for OneOfSchemaRewriter {
+    fn transform(&mut self, transform_schema: &mut schemars::Schema) {
+        // Apply this (self) transform to all subschemas
+        schemars::transform::transform_subschemas(self, transform_schema);
+
+        let mut schema: SchemaObject = match serde_json::from_value(transform_schema.clone().to_value()).ok()
+        {
+            // TODO (@NickLarsenNZ): At this point, we are seeing duplicate `title` when printing schema as json.
+            // This is because `title` is specified under both `extensions` and `other`.
+            Some(schema) => schema,
+            None => return,
+        };
+
+        hoist_one_of_enum_with_unit_variants(&mut schema);
+
+        // Convert back to schemars::Schema
+        if let Ok(schema) = serde_json::to_value(schema) {
+            if let Ok(transformed) = serde_json::from_value(schema) {
+                *transform_schema = transformed;
+            }
+        }
+    }
+}
+
+fn hoist_one_of_enum_with_unit_variants(kube_schema: &mut SchemaObject) {
+    // Run some initial checks in case there is nothing to do
+    let SchemaObject {
+        subschemas: Some(subschemas),
+        ..
+    } = kube_schema
+    else {
+        return;
+    };
+
+    let SubschemaValidation {
+        any_of: None,
+        one_of: Some(one_of),
+    } = subschemas.deref_mut()
+    else {
+        return;
+    };
+
+    if one_of.is_empty() {
+        return;
+    }
+
+    // At this point, we can be reasonably sure we need to hoist the oneOf
+    // subschema enums and types up to the schema level, and unset the oneOf field.
+    // From here, anything that looks wrong will panic instead of return.
+    // TODO (@NickLarsenNZ): Return errors instead of panicking, leave panicking up to the
+    // infallible schemars::Transform
+
+    // Prepare to ensure each variant schema has a type
+    let mut types = one_of.iter().map(|schema| match schema {
+        Schema::Object(SchemaObject {
+            instance_type: Some(r#type),
+            ..
+        }) => r#type,
+        Schema::Object(untyped) => panic!("oneOf variants need to define a type: {untyped:#?}"),
+        Schema::Bool(_) => panic!("oneOf variants can not be of type boolean"),
```

Comment:
is this a golang restriction or something related to how we have a weird `Schema` enum with a  "trivial boolean JSON schema" embedded? (which i tbh do not remember why we have)